lib/events.js

```diff
@@ -647,3 +652,84 @@ function once(emitter, name) {
     emitter.once(name, eventListener);
   });
 }
+
+const AsyncIteratorPrototype = ObjectGetPrototypeOf(
+  ObjectGetPrototypeOf(async function* () {}).prototype);
+
+function createIterResult(value, done) {
+  return { value, done };
+}
+
+function on(emitter, event) {
+  const unconsumedEvents = [];
+  const unconsumedPromises = [];
+  let error = null;
+
+  const iterator = ObjectSetPrototypeOf({
+    next() {
+      if (error) {
+        return PromiseReject(error);
+      }
+
+      const value = unconsumedEvents.shift();
+      if (value) {
+        return PromiseResolve(createIterResult(value, false));
+      }
+
+      return new Promise(function(resolve, reject) {
+        unconsumedPromises.push({ resolve, reject });
+      });
+    },
+
+    return() {
+      emitter.removeListener(event, eventHandler);
+      emitter.removeListener('error', errorHandler);
+
+      for (const promise of unconsumedPromises) {
```

Comment:
Since you return an iter result whose value is `true` we should set a boolean `finished = true` in this iterator.

Any future calls to `next()` can yield finished result or throw an InvalidStateException

Then the following behavior

```
const itr = on(emitter, 'what')
const r = await itr.next()
const r2 = await itr.return()
// r2.done === true
const r3 = await itr.next() // PENDING PERMANENTLY
```

Does not happen, instead `r3` would be `r3.done === true` or an Invalid exception.

Reply:
fixed.